pkg/sharky/shard.go

```diff
@@ -5,187 +5,140 @@
 package sharky
 
 import (
-	"encoding/json"
+	"context"
 	"os"
-	"sync"
 )
 
-// location models the location <shard, offset, length> of a chunk
+// Location models the location <shard, slot, length> of a chunk
 type Location struct {
 	Shard  uint8
-	Offset int64
-	Length int64
+	Slot   uint32
+	Length uint16
 }
 
-// operation models both read and write
-type operation struct {
-	location Location   // shard, offset, length combo - given for read, returned by write
-	buffer   []byte     // fixed size byte slice for data - copied to by read
-	data     []byte     // variable length byte slice - given for write
-	err      chan error // signal for end of operation
+// write models the input to a write operation
+type write struct {
+	buf []byte     // variable size read buffer
+	res chan entry // to put the result through
+}
+
+// entry models the output result of a write operation
+type entry struct {
+	loc Location // shard, slot, length combo
+	err error    // signal for end of operation
+}
+
+// read models the input to read operation (the output is an error)
+type read struct {
+	buf  []byte // variable size read buffer
+	slot uint32 // slot to read from
+	at   int    // within-chunk offset for partial reads
```

Comment:
i think that this is not needed for now. it is not hooked into the API, and it is very easy to add as a feature later on once we need it